klib/ntp.c

```diff
@@ -0,0 +1,158 @@
+#include <kernel.h>
+#include <lwip.h>
+
+#define NTP_SERVER  "pool.ntp.org"
+#define NTP_PORT    123
+
+#define NTP_EPOCH_DELTA 2208988800ul    /* Number of seconds between 1900 and 1970 */
+
+#define NTP_QUERY_INTERVAL  (10 * seconds(60))
+
+#define NTP_MAX_SLEW_RATE   0.0005  /* 500 PPM */
+
+struct ntp_ts {
+    u32 seconds;
+    u32 fraction;
+};
+
+struct ntp_packet {
+    u8 mode:3;
+    u8 vn:3;
+    u8 li:2;
+    u8 stratum;
+    u8 poll;
+    u8 precision;
+    u32 root_delay;
+    u32 root_dispersion;
+    u32 reference_id;
+    struct ntp_ts reference_ts;
+    struct ntp_ts originate_ts;
+    struct ntp_ts receive_ts;
+    struct ntp_ts transmit_ts;
+} __attribute((packed));
+
+declare_closure_struct(0, 1, void, ntp_query_func,
+    u64, overruns);
+
+static struct {
+    struct udp_pcb *pcb;
+    closure_struct(ntp_query_func, query_func);
+    timestamp last_raw;
+    err_t (*dns_gethostbyname)(const char *hostname, ip_addr_t *addr,
+            dns_found_callback found, void *callback_arg);
+    struct pbuf *(*pbuf_alloc)(pbuf_layer layer, u16_t length, pbuf_type type);
+    u8 (*pbuf_free)(struct pbuf *p);
+    err_t(*udp_sendto)(struct udp_pcb *pcb, struct pbuf *p,
+            const ip_addr_t *dst_ip, u16_t dst_port);
+    void (*runtime_memset)(u8 *a, u8 b, bytes len);
+    timestamp (*now)(clock_id id);
+    void (*clock_adjust)(timestamp now, double temp_cal, timestamp sync_complete, double cal);
+} ntp;
+
+static void timestamp_to_ntptime(timestamp t, struct ntp_ts *ntptime)
+{
+    ntptime->seconds = PP_HTONL(NTP_EPOCH_DELTA + sec_from_timestamp(t));
+    ntptime->fraction = PP_HTONL((u32)t);
+}
+
+static timestamp ntptime_to_timestamp(struct ntp_ts *ntptime)
+{
+    return (seconds(PP_NTOHL(ntptime->seconds) - NTP_EPOCH_DELTA) + PP_NTOHL(ntptime->fraction));
+}
+
+static s64 ntptime_diff(struct ntp_ts *t1, struct ntp_ts *t2)
+{
+    return (seconds(PP_NTOHL(t1->seconds) - PP_NTOHL(t2->seconds)) +
+            PP_NTOHL(t1->fraction) - PP_NTOHL(t2->fraction));
+}
+
+static void ntp_query(const ip_addr_t *server_addr)
+{
+    struct pbuf *p = ntp.pbuf_alloc(PBUF_TRANSPORT, sizeof(struct ntp_packet), PBUF_RAM);
+    if (p == 0)
+        return;
+    struct ntp_packet *pkt = p->payload;
+    ntp.runtime_memset(p->payload, 0, sizeof(*pkt));
+    pkt->vn = 3;    /* NTP version number */
+    pkt->mode = 3;  /* client mode */
+    timestamp_to_ntptime(ntp.now(CLOCK_ID_REALTIME), &pkt->transmit_ts);
+    ntp.udp_sendto(ntp.pcb, p, server_addr, NTP_PORT);
```

Comment:
The result of udp_sendto could be checked here, reporting to the user if the server cannot be contacted. Also, though the period of the update timer is long, it may be wise to either disable it or exponentially back off the period if NTP requests are not being completed after some threshold of attempts.

Reply:
I added an exponential backoff of the NTP query interval if NTP requests fail for NTP_QUERY_ATTEMPTS consecutive times (with NTP_QUERY_ATTEMPTS currently set to 8). Also, now errors are printed to the console.